src/mobu/services/business/notebookrunner.py

```diff
@@ -295,26 +315,88 @@ async def execute_code(self, session: JupyterLabSession) -> None:
             msg = f"Notebook {self._notebook.name} iteration {iteration}"
             self.logger.info(msg)
 
-            for cell in self.read_notebook(self._notebook):
-                code = "".join(cell["source"])
-                cell_id = cell.get("id") or cell["_index"]
-                ctx = CodeContext(
-                    notebook=self._notebook.name,
-                    path=str(self._notebook),
-                    cell=cell_id,
-                    cell_number=f"#{cell['_index']}",
-                    cell_source=code,
-                )
-                await self.execute_cell(session, code, cell_id, ctx)
-                if not await self.execution_idle():
-                    break
+            with self.timings.start(
+                "execute_notebook", self.annotations(self._notebook.name)
+            ) as sw:
+                try:
```

Comment:
This is not actionable feedback, just me musing, but something feels weird here. It feels like we're running timings and events in parallel, and are doing a lot of manual bookkeeping that creates deeply nested structure. My brain wants to find a new abstraction that unifies these layers. Something like... starting an event that can succeed or fail with a context manager, having it keep a top-level duration, and then possibly having it expose additional context managers that time specific subtasks, and then have the `__exit__` dunder method of the context manager capture success or failure based on whether an exception was thrown?

Happy to merge this without that change, but I feel like we're getting a bit of complexity creep here. Maybe a bunch of it goes away when we throw out the old way of recording timings?